### DISCUSSION TITLE
How to pass custom attributes to the registration form?

### DISCUSSION THREAD

**Discussion by mkreis** [keycloak/keycloak#9449] [Q&A] [answered]
Currently everybody can register for our app using an URL posted on different websites.

But the URL is the same on all pages, thus it is not possible to know where an user originated from after the sign-up.
Therefore I was thinking about using custom URLs for the registration page (with additional GET parameters like "&custom_attribute_reference=landingpage1") and store this parameter as an custom attribute on the user.

Is this possible? Or is there some other simple way to solve this problem?

Any help or pointers are appreciated!

Btw. someone asked a similar question already here https://stackoverflow.com/questions/58005895/is-it-possible-to-add-custom-get-parameters-in-keycloak-registration-page but unfortunately it was never answered.

Thanks!

**Comment by bs-matil** [comment, marked as answer]
You could add a authenticator to your authentication flow for registration before the registration step, which reads additional get parameters from the request to the authentication session. After the user successfully registers another authenticator which follows the registration authenticator can persist this attribute with the user or somewhere else.

**Reply by mkreis** [reply]
Hi Markus, thank you very much, with your response I was able to solve the problem.

For others looking for the same solution, here is what I did:

1. Create a new .jar file with the following files:

a) META-INF/keycloak-scripts.json
```
{
  "authenticators": [
    {
      "name": "Store reference in session",
      "fileName": "ref2session.js",
      "description": "Store reference query param in session"
    },
    {
      "name": "Set reference as attribute",
      "fileName": "session2attrib.js",
      "description": "Store reference from sessin as custom attribute"
    }
  ],
  "policies": [],
  "mappers": []
}
```

b) ref2session.js
```
function authenticate(context) {
    var session = context.getAuthenticationSession();
    var ref = context.getUriInfo().getQueryParameters(false).getFirst("reference");
    if(ref) session.setUserSessionNote("reference", ref);
    context.success();
}
```

c) session2attrib.js
```
function authenticate(context) {
    var session = context.getAuthenticationSession();
    var ref = session.getUserSessionNotes().get("reference");
    if(user && ref) user.setSingleAttribute("reference", ref);
    context.success();
}
```

2. Place this file in folder `keycloak/standalone/deployments`
3. Add JVM Option `-Dkeycloak.profile.feature.scripts=enabled`
4. Restart the server
5. In the Admin console, go to Authentication and make a copy of authentication flow "Registration"
6. In the new flow, click on Add Execution and select  "Store reference in session"
7. Using the arrow buttons, move the new entry to the top and select "Required"
8. click on Add Execution and select "Set reference as attribute"
9. Leave this line on the bottom and select "Required" as well
9. Now you can use a link to the registration page like this:
`https://<your_host>/auth/realms/<your_realm>/protocol/openid-connect/registrations?client_id=<your_client>&response_type=code&scope=openid%20email&redirect_uri=<your_host>&reference=<some_reference_value>`
Where any value for the parameter `reference` is stored as a custom attribute on the new user.

**Reply by bs-matil** [reply]
One Note on this. Your choice to use JS authenticators is reasonable as its fast way forward. You should know that this feature seems to be in preview for a long time and is not well supported. 
For a long term solution I would recommend to use a Java SPI implementation.

**Reply by mkreis** [reply]
Okay, thanks for the heads up!

**Reply by NicholasMantovani** [reply]
@bs-matil hello can you provide an example with the SPI implementation?

**Reply by javiercanas2023** [reply]
hello can you provide an example with the SPI implementation?

**Reply by hadpro24** [reply]
I am also still wait this for SPI implementation

**Reply by SeanCodeMedia** [reply]
Can someone please answer this ?

**Comment by javiercanas2023** [comment]
I'm developing for delivery projects. I don't know what an ISP is. If you'd like to tell me more, please write to me.


________________________________
